stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/composer/MessageComposerView.kt

```diff
@@ -0,0 +1,195 @@
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.launch
+
+@ExperimentalStreamChatApi
+public class MessageComposerView : ConstraintLayout {
+
+    private lateinit var binding: StreamUiMessageComposerBinding
+
+    /**
+     * @property messageInputState The current state of the [MessageComposerView] distributed using [Flow] API.
+     *
+     * Whenever the value of this property changes updated the [MessageComposerView] re-renders automatically.
+     * If you are using custom [View] implementation for [leadingContent], [centerContent], or [trailingContent] you may
+     * want to collect [messageInputState] items to update them when the state changes.
+     */
+    public val messageInputState: MutableStateFlow<MessageInputState> = MutableStateFlow(MessageInputState())
+
+    /**
+     * @property onSendMessageAction Handler invoked when the user taps on the send message button.
+     */
+    public var onSendMessageAction: () -> Unit = {}
+
+    /**
+     * @property onCancelAction Handler invoked when the user cancels the active action (in Reply or Edit modes).
+     */
+    public var onCancelAction: () -> Unit = {}
+
+    /**
+     * @property onInputTextChanged Handler invoked when the user enters text in message input field.
+     */
+    public var onInputTextChanged: (String) -> Unit = {}
+
+    private val defaultLeadingContent: View by lazy {
+        val binding = StreamUiMessageComposerDefaultLeadingContentBinding.inflate(
+            streamThemeInflater,
+            binding.trailingContent,
+            false
+        )
+        binding.attachmentsButton.isVisible = false
+        binding.commandsButton.isVisible = false
+        binding.root
+    }
+
+    /**
+     * @property leadingContent A function returning content visible on the left.
+     */
+    public var leadingContent: MessageComposerView.() -> View = {
+        defaultLeadingContent
+    }
+        set(value) {
+            field = value
+            binding.leadingContent.apply {
+                removeAllViews()
+                addView(value())
+            }
+        }
+
+    private val defaultCenterContent: View by lazy {
+        val binding = StreamUiMessageComposerDefaultCenterContentBinding.inflate(
+            streamThemeInflater,
+            binding.trailingContent,
+            false
+        )
+        binding.messageEditText.addTextChangedListener(object : TextWatcherAdapter() {
+            override fun afterTextChanged(s: Editable) {
+                onInputTextChanged(s.toString())
+            }
+        })
+        binding.clearCommandButton.isVisible = false
+        binding.selectedAttachmentsContainer.isVisible = false
+        binding.commandBadge.isVisible = false
+        binding.messageReplyView.isVisible = false
+        CoroutineScope(Dispatchers.Main).launch {
```

Comment:
Why do we collect the state here, inside the default content? This seems like more of an `init()` block. In this case, if the user doesn't collect the state properly if they override the UI, then it won't update if I understood it correctly

Also - do we cancel this dispatcher if the user replaces the content? We don't have access to the Scope or the Job to call cancel properly.